### PR TITLE
Return unsafe sharing error from backend

### PR DESCRIPTION
**Pull request by rauchy**
## What type of PR is this? (check all applicable)
- [x] Refactor

## Description
Create a message explaining that unsafe queries can't be shared in the backend, and use that message in public dashboards and embeds.

## Related Tickets & Documents
Replaces the behavior introduced in #3659 and #3988 (/cc @ranbena)

### CODE REVIEW

File: redash/handlers/query_results.py [hunk 1]
```diff
@@ -182,7 +182,12 @@ def post(self, query_id):
         if has_access(query, self.current_user, allow_executing_with_view_only_permissions):
             return run_query(query.parameterized, parameter_values, query.data_source, query_id, max_age)
         else:
-            return {'job': {'status': 4, 'error': 'You do not have permission to run queries with this data source.'}}, 403
+            if current_user.is_api_user() and not query.parameterized.is_safe:
```
Comment (arikfr) [inline redash/handlers/query_results.py]: Tests?

File: redash/handlers/query_results.py [hunk 2]
```diff
@@ -182,7 +182,12 @@ def post(self, query_id):
...
+                message = 'This query contains potentially unsafe parameters and cannot be executed on a publicly shared resource.'
```
Comment (arikfr) [inline redash/handlers/query_results.py]:
```suggestion
                message = 'This query contains potentially unsafe parameters and cannot be executed on a shared dashboard or visualization embed.'
```

Comment (rauchy) [reply redash/handlers/query_results.py]: I deliberately used "resource" to avoid the word "dashboard" appearing on embeds and vice versa, but 🤷🏻‍♂️

Comment (arikfr) [reply redash/handlers/query_results.py]: I know, but "publicly shared resource" sounds like something taken from a programming manual and not very friendly. :)

File: redash/handlers/query_results.py (hunk 1 above)
Comment (rauchy) [reply redash/handlers/query_results.py]: 👍done in 474dae1

File: redash/handlers/query_results.py (hunk 2 above)
Comment (rauchy) [reply redash/handlers/query_results.py]: image

File: redash/handlers/query_results.py [hunk 3]
```diff
@@ -15,8 +15,16 @@
...
 
 
-def error_response(message):
-    return {'job': {'status': 4, 'error': message}}, 400
+def error_response(message, http_status=400):
+    return {'job': {'status': 4, 'error': message}}, http_status
+
+
+error_messages = {
+    'unsafe_when_shared': error_response('This query contains potentially unsafe parameters and cannot be executed on a shared dashboard or visualization embed.', 403),
```
Comment (ranbena) [inline redash/handlers/query_results.py]:
```suggestion
    'unsafe_when_shared': error_response('This query contains potentially unsafe parameters and cannot be executed on a shared dashboard or an embedded visualization.', 403),
```

File: redash/handlers/query_results.py [hunk 4]
```diff
@@ -15,8 +15,16 @@
...
 
-def error_response(message):
-    return {'job': {'status': 4, 'error': message}}, 400
+def error_response(message, http_status=400):
+    return {'job': {'status': 4, 'error': message}}, http_status
+
+
+error_messages = {
+    'unsafe_when_shared': error_response('This query contains potentially unsafe parameters and cannot be executed on a shared dashboard or an embedded visualization.', 403),
+    'unsafe_on_view_only': error_response('This query contains potentially unsafe parameters and cannot be executed on this data source.', 403),
```
Comment (arikfr) [inline redash/handlers/query_results.py]:
```suggestion
    'unsafe_on_view_only': error_response('This query contains potentially unsafe parameters and cannot be executed with read-only access to this data source.', 403),
```

Maybe something like this?

**Pull request merged**